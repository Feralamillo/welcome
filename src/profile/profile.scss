@import 'src/styles/variables.scss';

.container {
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100vh;
}

.user-profile {
  background-color: $white;
  border-radius: 10px;
  box-shadow: 0px 4px 4px rgba($black, 0.25);
  text-align: center;
  padding: 2rem;

  .user-image {
    height: 200px;
    width: 200px;
    border-radius: 50%;
  }

  .title {
    color: $primary;
  }

  .food {
    font-weight: bold;
    color: $primary;
  }
}
