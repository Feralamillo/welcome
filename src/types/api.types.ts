export interface SuccessResponse<T> {
  success: true;
  result: T;
}

export interface FailedResponse<T> {
  success: false;
  status: number;
  message?: string;
}

export type ApiResponse<T> = SuccessResponse<T> | FailedResponse<T>;
