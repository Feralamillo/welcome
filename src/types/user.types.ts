export interface User {
  firstName: string;
  lastName: string;
  img: string;
  food: string;
}
