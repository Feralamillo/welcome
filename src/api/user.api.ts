import axios from 'axios';

import { ApiResponse } from '../types/api.types';
import { User } from '../types/user.types';

export const userApi = {
  getUser: async (): Promise<ApiResponse<User>> => {
    try {
      /* @TODO: API_URL as env variable */
      const response = await axios.get('http://localhost:4000/api/user');

      return {
        success: true,
        result: response.data,
      };
    } catch (err) {
      if (err.response) {
        return {
          success: false,
          status: err.response.status,
        };
      }

      return {
        success: false,
        status: 404,
      };
    }
  },
};
