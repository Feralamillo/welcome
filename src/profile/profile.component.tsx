import React from 'react';

import { User } from '../types/user.types';

interface Props {
  user: User;
}

export const ProfileComponent: React.FC<Props> = ({
  user: { firstName, lastName, img, food },
}) => (
  <div className='container'>
    <div className='user-profile'>
      <img src={img} alt={`${firstName} ${lastName}`} className='user-image' />
      <h1 className='title'>{`${firstName} ${lastName}`}</h1>
      <p>
        My favourite food is <span className='food'>{food}</span>
      </p>
    </div>
  </div>
);
