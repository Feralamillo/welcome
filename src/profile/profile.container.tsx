import React, { useEffect, useState } from 'react';

import { userApi } from '../api/user.api';
import { User } from '../types/user.types';
import { Loader } from '../ui-components/loader/loader';
import { ProfileComponent } from './profile.component';

import './profile.scss';

export const Profile: React.FC = () => {
  const [user, setUser] = useState<User | null>(null);
  const [isFetching, setFetching] = useState(true);

  const fetchUser = async () => {
    const response = await userApi.getUser();

    if (response.success) {
      setUser(response.result);
    } else {
      /* @TODO: error management */
      setUser(null);
    }

    setFetching(false);
  };

  useEffect(() => {
    fetchUser();
  }, []);

  if (isFetching) {
    return (
      <div className='container'>
        <Loader />
      </div>
    );
  }

  if (!user) {
    return (
      <div className='container'>
        <div>Sorry...there has been an error</div>
      </div>
    );
  }

  return <ProfileComponent user={user} />;
};
