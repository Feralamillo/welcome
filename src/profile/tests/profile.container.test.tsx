import { render } from '@testing-library/react';
import { userApi } from '../../api/user.api';
import { User } from '../../types/user.types';

import { Profile } from '../profile.container';

const getUserInfo = jest.spyOn(userApi, 'getUser');

jest.mock('../profile.component', () => ({
  ProfileComponent: 'mock-profile-component',
}));
jest.mock('../../ui-components/loader/loader', () => ({
  Loader: 'mock-loader',
}));

describe('<Profile>', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should render with data', (done) => {
    expect.assertions(1);
    const mockUser: User = {
      firstName: 'Jim',
      lastName: 'Morrison',
      food: 'Pizza',
      img: 'srcImage',
    };
    getUserInfo.mockResolvedValueOnce({
      success: true,
      result: mockUser,
    });

    const { container } = render(<Profile />);

    setTimeout(async () => {
      expect(container).toMatchSnapshot();
      done();
    });
  });

  it('should render with error', (done) => {
    expect.assertions(1);
    getUserInfo.mockResolvedValueOnce({
      success: false,
      status: 400,
    });

    const { container } = render(<Profile />);

    setTimeout(async () => {
      expect(container).toMatchSnapshot();
      done();
    });
  });

  it('should render loading', () => {
    expect.assertions(1);
    const { container } = render(<Profile />);

    expect(container).toMatchSnapshot();
  });
});
