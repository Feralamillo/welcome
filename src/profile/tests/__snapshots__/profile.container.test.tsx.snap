// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile> should render loading 1`] = `
<div>
  <div
    class="container"
  >
    <mock-loader />
  </div>
</div>
`;

exports[`<Profile> should render with data 1`] = `
<div>
  <mock-profile-component
    user="[object Object]"
  />
</div>
`;

exports[`<Profile> should render with error 1`] = `
<div>
  <div
    class="container"
  >
    <div>
      Sorry...there has been an error
    </div>
  </div>
</div>
`;
